### ISSUE TITLE
In evidenceToMatrix(dda_pip, return_EList = TRUE)

### ISSUE THREAD

**Issue by EidrianGM** [DavisLaboratory/msImpute#5]
When applying the basic functions:

```r
dda_pip <- mspip("path/to/MaxQuant/Output", skip_weights = TRUE)
y_pip <- evidenceToMatrix(dda_pip, return_EList = TRUE)
``` 
> Error in genes[, peptide_id] : incorrect number of dimensions
> y_pip <- evidenceToMatrix(dda_pip, return_EList = TRUE)

I think this is happening because I only have the column "PeptideID" in my dda_pip and the line 18 of the function _evidenceToMatrix_ 

```r    
genes <- evidence[, evidence_colnames %in% tolower(meta_attrs)]
```
returns an uni-dimensional character vector.

A solution would be to substitute next line, 19, with:
```r    
if (is.null(dim(genes))){
  genes <- genes[!duplicated(genes)]
  genes <- genes[match(rownames(E), genes)]
}else{
  genes <- genes[!duplicated(genes), ]
  genes <- genes[match(rownames(E), genes[, peptide_id]), ]
}
```

My issue is that I am not sure if this is the proper way to solve this or should the genes variable be two dimensional from the very beggining?

**Comment by soroorh** [comment]
Hi Eidrian.

`genes` is meant to be a data frame since it makes up the gene component of the Elist obj (line 73)
```
new("EListRaw", list(E=E, genes = genes))
```
you can specify the corresponding peptideID key in your `dda_pip` through `peptide_id` argument to `evidenceToMatrix`.

Hope this helps

**Comment by EidrianGM** [comment]
That is what I was suspecting... `peptide_id` is defined as default `"PeptideID"` which is actually present in the `dda_pip` dataframe, thus changing this will not solve the issue. 

I am not sure but I think there should be some type of problem either in the parser or in the MaxQuant files format that I have. Not only because of the first mentioned issue, but also because the function `evidenceToMatrix` is expecting meta-attributes (line 21-23) 
```r
meta_attrs <- c(peptide_id, "Sequence", "Length", "Modifications", 
                    "Modified.sequence", "Leading.razor.protein", "Gene.Names", 
                    "Protein.Names", "Charge")
```
but actually my resulting `dda_pip` is returned only with three columns
```r
> colnames(dda_pip)
[1] "Raw.file"  "PeptideID" "Intensity"
```
Nevertheless, my MaxQuant _evidence.txt_ has all those expected attributes excepting "Gene.Names" and "Protein.Names".


Another things that looks weird to me is that `dda_pip$PeptideID` shows peptides like these:
```r
468  _AIEM(Oxidation (M))EGLTWGAHQFIPIGFGIK_3                _AIEM(Oxidation (M))EGLTWGAHQFIPIGFGIK_3
469  _AIEMEGLTWGAHQFIPIGFGIK_2                               _AIEMEGLTWGAHQFIPIGFGIK_2
```
There seems to be a ```paste(colA,colB,colC,sep='_') ``` where some columns are empty or not found. Maybe the protein name?

**Comment by soroorh** [comment]
thanks Adrian.

The wrappers were written assuming specific (most likely old) versions of maxquant. It might be that there are very minor differences in attribute names in your evidence.txt . Is it possible to please provide the first few (6-10) lines of your file including the header line?

by default PeptideID, if generated by msImpute, is the concatenation of modified sequence and charge state.

Hope this helps,
S

**Comment by EidrianGM** [comment]
Yes! Thanks so much for your quick response here you have a head subset of the file:
[headSubset_evidence.txt](https://github.com/DavisLaboratory/msImpute/files/14922318/headSubset_evidence.txt)

**Comment by EidrianGM** [comment]
Okey, I run manually 
```r
dda_pip <- mspip("path/to/MaxQuant/Output", skip_weights = TRUE)
```
and it was all about the `skip_weights = TRUE`. This parameter set to TRUE return a three columns dataframe as shown in my previous comment. Which only have one column in common with the `_meta_attrs_` in `evidenceToMatrix`, "PeptideID", and returns genes as an uni-dimensional character vector (see first comment). In order to avoid genes being initially defined as a character vector, you can use  the solution posted innitially, or simply replace lines 18 and 19 with:

```r
genes <- evidence[, evidence_colnames %in% tolower(meta_attrs), drop=F]
genes <- genes[!duplicated(genes), ,drop=F]
# Setting drop=F maintains the dimensions of the DF
```
These solutions allows to use both `skip_weights = TRUE` or `skip_weights = FALSE` in `mspip` function.

When `skip_weights = FALSE` the dataframe returned by mspip has more columns in common with the `meta_attrs` of `evidenceToMatrix` function and do not causes the issue with genes being returned as an uni-dimensional character vector.

**Comment by soroorh** [comment]
great - thanks for suggestions. Will have a look at mspip with different skip_weights and incorporate suggestions.
